zebra/zebra_srv6_vty.c

```diff
@@ -271,35 +275,42 @@ DEFUN (no_srv6_locator,
 
 DEFPY (locator_prefix,
        locator_prefix_cmd,
-       "prefix X:X::X:X/M$prefix [func-bits (16-64)$func_bit_len]",
+       "prefix X:X::X:X/M$prefix [func-bits (16-64)$func_bit_len] \
+	       [block-len (16-64)$block_bit_len] [node-len (16-64)$node_bit_len]",
        "Configure SRv6 locator prefix\n"
        "Specify SRv6 locator prefix\n"
        "Configure SRv6 locator function length in bits\n"
-       "Specify SRv6 locator function length in bits\n")
+       "Specify SRv6 locator function length in bits\n"
+       "Configure SRv6 locator block length in bits\n"
+       "Specify SRv6 locator block length in bits\n"
+       "Configure SRv6 locator node length in bits\n"
+       "Specify SRv6 locator node length in bits\n")
 {
 	VTY_DECLVAR_CONTEXT(srv6_locator, locator);
 	struct srv6_locator_chunk *chunk = NULL;
 	struct listnode *node = NULL;
 
 	locator->prefix = *prefix;
 
-	/*
-	 * TODO(slankdev): please support variable node-bit-length.
-	 * In draft-ietf-bess-srv6-services-05#section-3.2.1.
-	 * Locator block length and Locator node length are defined.
-	 * Which are defined as "locator-len == block-len + node-len".
-	 * In current implementation, node bits length is hardcoded as 24.
-	 * It should be supported various val.
-	 *
-	 * Cisco IOS-XR support only following pattern.
-	 *  (1) Teh locator length should be 64-bits long.
-	 *  (2) The SID block portion (MSBs) cannot exceed 40 bits.
-	 *      If this value is less than 40 bits,
-	 *      user should use a pattern of zeros as a filler.
-	 *  (3) The Node Id portion (LSBs) cannot exceed 24 bits.
-	 */
-	locator->block_bits_length = prefix->prefixlen - 24;
-	locator->node_bits_length = 24;
+	/* Resolve optional arguments */
```

Comment:
Should we do some validation here to ensure that "block_len + node_len + function_len + arg_len" is not > 128 bits? 

Consider the following scenario:

"locator r2
    prefix cafe:3:3::1/128 func-bits 16
   exit"

In case someone configures the above locator (which is a wrong config), then the SID parameters will look something like the following: 
 
  block_bits_length = 104 'h',
  node_bits_length = 24 '\030',
  function_bits_length = 16 '\020',
  argument_bits_length = 0 '\000',
===> Total length is 144!

Here block_len + node_len + function_len + arg_len exceeds 128 bits and BGP will end up sending an update with L3 srv6 sub-sub tlv containing transpose offset = 128(block_len + node_len) and transpose_len = 16 which is wrong.

according to section 3.1 of RFC 8986:
 "This document defines an SRv6 SID as consisting of LOC:FUNCT:ARG,
   where a locator (LOC) is encoded in the L most significant bits of
   the SID, followed by F bits of function (FUNCT) and A bits of
   arguments (ARG).  L, the locator length, is flexible, and an operator
   is free to use the locator length of their choice.  F and A may be
   any value as long as L+F+A <= 128"

Reply:
I agree. I updated the PR to add a check to validate the locator (commit https://github.com/FRRouting/frr/pull/11673/commits/983d79618bbd2f8c247b1cd2cbfd3b8b6d6a2fce). We accept a locator only if ``block_len + node_len + function_len + arg_len <= 128``.